test/wpool_process_SUITE.erl

```diff
@@ -104,24 +106,87 @@ info(_Config) ->
 cast(_Config) ->
   {ok, Pid} = wpool_process:start_link(?MODULE, echo_server, {ok, state}, []),
   wpool_process:cast(Pid, {noreply, newstate}),
-  newstate = wpool_process:call(?MODULE, state, 5000),
+  newstate = get_state(?MODULE),
   wpool_process:cast(Pid, {noreply, newerstate, 0}),
   timer:sleep(100),
-  timeout = wpool_process:call(?MODULE, state, 5000),
+  timeout = get_state(?MODULE),
   wpool_process:cast(Pid, {stop, normal, state}),
   timer:sleep(1000),
   false = erlang:is_process_alive(Pid),
 
   {comment, []}.
 
+-spec continue(config()) -> {comment, []}.
+continue(_Config) ->
+  C = fun(ContinueState) -> {noreply, ContinueState} end,
+  %% init/1 returns {continue, continue_state}
+  {ok, Pid} =
+    wpool_process:start_link(
+      ?MODULE, echo_server, {ok, state, {continue, C(continue_state)}}, []),
+  continue_state = get_state(Pid),
+
+  %% handle_call/3 returns {continue, ...}
+  ok = wpool_process:call(Pid, {reply, ok, state, {continue, C(continue_state_2)}}, 5000),
+  continue_state_2 = get_state(Pid),
+  try wpool_process:call(Pid, {noreply, state, {continue, C(continue_state_3)}}, 100) of
+    Result -> ct:fail("Unexpected Result: ~p", [Result])
+  catch
+    _:{timeout, _} ->
+      continue_state_3 = get_state(Pid)
+  end,
+
+  %% handle_cast/2 returns {continue, ...}
+  wpool_process:cast(Pid, {noreply, state, {continue, C(continue_state_4)}}),
+  continue_state_4 = get_state(Pid),
+
+  %% handle_continue/2 returns {continue, ...}
+  SecondContinueResponse = C(continue_state_5),
+  FirstContinueResponse = {noreply, another_state, {continue, SecondContinueResponse}},
+  CastResponse = {noreply, state, {continue, FirstContinueResponse}},
+  wpool_process:cast(Pid, CastResponse),
+  continue_state_5 = get_state(Pid),
+
+  %% handle_info/2 returns {continue, ...}
+  Pid ! {noreply, state, {continue, C(continue_state_6)}},
+  continue_state_6 = get_state(Pid),
+
+  %% handle_continue/2 returns {continue, ...}
+  SecondContinueResponse = C(continue_state_5),
+  FirstContinueResponse = {noreply, another_state, {continue, SecondContinueResponse}},
+  CastResponse = {noreply, state, {continue, FirstContinueResponse}},
+  wpool_process:cast(Pid, CastResponse),
+  continue_state_5 = get_state(Pid),
+
+  %% handle_continue/2 returns timeout = 0
+  wpool_process:cast(Pid, {noreply, state, {continue, {noreply, continue_state_7, 0}}}),
+  timer:sleep(100),
```

Comment:
From my experience sleeping arbitrary time may easily fail on hosted CIs like travis. What we recently try to do in esl/MongooseIM is a busy wait with retries. In our tests we have a function [wait_until/2 or /4](https://github.com/esl/MongooseIM/blob/7ea42e809b0e4eb77b8df1342e9c434f1047b67b/big_tests/tests/mongoose_helper.erl#L275) which takes a fun to execute and expected value. The fun is called and `wait_until` returns when the result matches expected value. If not, the `wait_until` function waits some time (by default 100ms) and retries until a deadline pass (by default 5s).

This is just a comment not a request for changes.

Reply:
And here's the random fail on travis due to waiting to short (as far as I can tell).

```
%%% wpool_process_SUITE ==> info: FAILED
%%% wpool_process_SUITE ==> {{badmatch,newerstate},
 [{wpool_process_SUITE,info,1,
                       [{file,"/home/travis/build/inaka/worker_pool/_build/test/lib/worker_pool/test/wpool_process_SUITE.erl"},
                        {line,96}]},
  {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
  {test_server,run_test_case_eval1,6,[{file,"test_server.erl"},{line,1045}]},
  {test_server,run_test_case_eval,9,[{file,"test_server.erl"},{line,977}]}]}
```

This happened to my PR #145. When I restarted the job, it worked just fine.

Reply:
Created #146 for that :)